src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+
+    return req;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_ptr_t
+ucp_stream_recv_request_put(ucp_request_t *req, size_t *length,
+                            ucs_status_t status)
+{
+    *length = req->recv.state.offset;
+    ucp_request_put(req);
+    return UCS_STATUS_PTR(status);
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
+                 (ep, buffer, count, datatype, cb, flags),
+                 ucp_ep_h ep, void *buffer, size_t *count,
+                 ucp_datatype_t datatype, ucp_stream_recv_callback_t cb,
+                 unsigned flags)
+{
+    ucp_ep_ext_stream_t *ep_stream;
+    ucp_request_t       *req;
+    ucs_status_ptr_t    rdesc;
+    ucs_status_t        status;
+
+    if (UCP_DT_IS_GENERIC(datatype)) {
+        ucs_error("ucp_stream_recv_nb doesn't support generic datatype");
+        return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+    }
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    req = ucp_stream_recv_request_get(ep->worker, buffer, *count, datatype, cb);
+    if (ucs_unlikely(req == NULL)) {
+        req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ep_stream = ep->ext.stream;
+
+    /* if there are already posted requests, need to queue this one */
+    if (!ucs_queue_is_empty(&ep_stream->reqs)) {
+        ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+        goto out;
+    }
+
+    /* OK, lets obtain all arrived data which matches the recv size */
+    while (req->recv.state.offset < req->recv.length) {
+        rdesc = ucp_stream_recv_data_nb_internal(ep_stream);
+        if (rdesc == NULL) {
+            /* No data any more. Enqueue empty request or complete if some
+             * data is filled */
+            if (req->recv.state.offset == 0) {
+                ucs_queue_push(&ep_stream->reqs, &req->recv.queue);
+            } else {
+                req = ucp_stream_recv_request_put(req, count, UCS_OK);
+            }
+            goto out;
+        }
+
+        status = ucp_stream_process_rdesc(rdesc, ep_stream, req);
+        if (ucs_unlikely(status != UCS_OK)) {
+            req = ucp_stream_recv_request_put(req, count, status);
+            goto out;
+        }
+    }
+
+    ucs_assert(req->recv.state.offset == req->recv.length);
+    req = ucp_stream_recv_request_put(req, count, UCS_OK);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+    return UCS_PTR_IS_PTR(req) ? (req + 1) : req;
+}
+
+static UCS_F_ALWAYS_INLINE ucp_recv_desc_t *
+ucp_stream_am_rdesc_get(ucp_worker_t *worker, void *data, size_t length,
+                        unsigned am_flags)
+{
+    ucp_recv_desc_t *rdesc;
 
     if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
         /* slowpath */
         rdesc        = (ucp_recv_desc_t *)data - 1;
         rdesc->flags = UCP_RECV_DESC_FLAG_UCT_DESC;
-        status       = UCS_INPROGRESS;
     } else {
-        status = UCS_OK;
+        /* TODO: optimize if there are posted recvs */
```

Comment:
i think we should do it right from the 1st time, so pls optimize this as part of current PR

Reply:
it adds a lot of overhead with testing/development on other branches. Again, i would do all optimizations after all functionality and testing PRs are merged.

Reply:
well, since this is not a just small optimization, but a change in the logic and potentially data structures, better to do it now before merging. much harder to clean the code after it's already merged.

Reply:
Much harder to have several opened PRs with cross dependencies and dev branch at the same time instead of to do the things incrementally.